src/pages/index.js

```diff
@@ -45,40 +89,106 @@ addCardMOB.addEventListener("click", () => {
   addPlacePopup.open();
 });
 
+profilePic.addEventListener("click", () => {
+  profilePicPopup.open();
+});
+
 //***************************************************************************************************************************//
 //                                             Event Handlers                                                                //
 //***************************************************************************************************************************//
 
 function handleEditProfile(userData) {
+  const submitButton = editProfileModal.querySelector(
+    "#edit-profile-modal-submit-button"
+  );
+  submitButton.textContent = "Saving...";
+  api.patchProfileInfo(userData.name, userData.job);
   userInfo.setUserInfo(userData);
   editProfilePopup.popupForm.reset();
   editProfileFormValidator.toggleButtonState();
   editProfilePopup.close();
+  submitButton.textContent = "Save";
 }
 
 function handleAddCardSubmit(cardData) {
-  createCard(cardData, "prepend");
-  addPlacePopup.popupForm.reset();
-  addPlaceFormValidator.toggleButtonState();
-  addPlacePopup.close();
+  const submitButton = addCardModal.querySelector(
+    "#add-card-modal-submit-button"
+  );
+  submitButton.textContent = "Saving...";
+
+  api
+    .addNewCard(cardData)
+    .then((data) => {
+      addNewCard(data, "prepend");
+      addPlacePopup.popupForm.reset();
+      addPlaceFormValidator.toggleButtonState();
+      addPlacePopup.close();
+      submitButton.textContent = "Create";
+    })
+    .catch((error) => {
+      console.error("Error adding card:", error);
+    });
+}
+
+let cardtoDeleteData;
+
+function handleConfirmDelete(data) {
+  confirmDeletePopup.open();
+  cardtoDeleteData = data;
+}
+
+function handleDeleteCard() {
+  api.deleteCard(cardtoDeleteData._id);
+  const image = document.querySelector(`img[src='${cardtoDeleteData.link}'`);
+  const deleteCard = image.closest(".card");
+  deleteCard.remove();
+  confirmDeletePopup.close();
+}
+
+function handleEditPicSubmit({ link }) {
+  const submitButton = editProfilePicModal.querySelector(
+    "#edit-pic-modal-submit-button"
+  );
+  submitButton.textContent = "Saving...";
```

Comment:
**Could be improved**


If it’s interesting for you here is how we can make a universal function for handling any submit. We can get rid of such duplicating as **loading effect**, **closing** and **catching errors**


```js

// You can make a universal function that accepts a request, popup instance and optional loading text  
function handleSubmit(request, popupInstance, loadingText = "Saving...") {

  // here we change the button text
  popupInstance.renderLoading(true, loadingText);
  request()
    .then(() => {
      // We need to close only in `then`
      popupInstance.close()
    })
      // we need to catch possible errors
      // console.error is used to handle errors if you don’t have any other ways for that
     .catch(console.error)
    // in `finally` we need to return the initial button text back in any case
    .finally(() => {
      popupInstance.renderLoading(false);
    });
}

```


You need to pass the **request** into the call of `handleSubmit` like that:



```js

// here is an example of the profile form handling
function handleProfileFormSubmit(inputValues) {
  // we create a function that returns a promise
  function makeRequest() {
    // `return` lets us use a promise chain `then, catch, finally` inside `handleSubmit`
    return api.editProfile(inputValues).then((userData) => {
      userInfo.setUserInfo(userData)
    });
  }
  // Here we call the function passing the request, popup instance and if we need some other loading text we can pass it as the 3rd argument
  handleSubmit(makeRequest, profilePopup);
}



```